KEP-1432: AdmissionCheck per ResourceFlavor (#1935)
kubernetes-sigs/kueue: keps/1432-admission-check-per-resource-flavor/README.md

Reviewer: please reference kep-993

# KEP-1432: AdmissionChecks per ResouceFlavor

<!--
This is the title of your KEP. Keep it short, simple, and descriptive. A good
title can help communicate what the KEP is and should be considered as part of
any review.
-->

<!--
A table of contents is helpful for quickly jumping to sections of a KEP and for
highlighting any additional information provided beyond the standard KEP
template.

Ensure the TOC is wrapped with
  <code>&lt;!-- toc --&rt;&lt;!-- /toc --&rt;</code>
tags, and then generate with `hack/update-toc.sh`.
-->

<!-- toc -->
- [Summary](#summary)
- [Motivation](#motivation)
  - [Goals](#goals)
  - [Non-Goals](#non-goals)
- [Proposal](#proposal)
  - [User Stories (Optional)](#user-stories-optional)
    - [Story 1](#story-1)
  - [Notes/Constraints/Caveats (Optional)](#notesconstraintscaveats-optional)
  - [Risks and Mitigations](#risks-and-mitigations)
- [Design Details](#design-details)
  - [Test Plan](#test-plan)
    - [Unit Tests](#unit-tests)
    - [Integration tests](#integration-tests)
- [Drawbacks](#drawbacks)
- [Alternatives](#alternatives)
<!-- /toc -->

## Summary

This KEP introduces assigning AdmissionChecks to a ResourceFlavor.

<!--

> This section is incredibly important for producing high-quality, user-focused

Reviewer 2: Please remove comments

documentation such as release notes or a development roadmap. It should be
possible to collect this information before implementation begins, in order to
avoid requiring implementors to split their attention between writing release
notes and implementing the feature itself. KEP editors and SIG Docs
should help to ensure that the tone and content of the `Summary` section is
useful for a wide audience.

A good summary is probably at least a paragraph in length.

Both in this section and below, follow the guidelines of the [documentation
style guide]. In particular, wrap lines to a reasonable length, to make it
easier for reviewers to cite specific portions, and to minimize diff churn on
updates.

[documentation style guide]: https://github.com/kubernetes/community/blob/master/contributors/guide/style-guide.md
-->

## Motivation

Currently, AdmissionChecks are only assigned to a ClusterQueue. This results in running `AdmissionChecks` regardless of
the ResourceFlavor used by a Workload, even when its unnecessary. We would like to have an ability to assign AdmissionChecks

> to a specific ResourceFlavor, so the whole mechanism is more flexible and efficient.

Reviewer 2: It's not that much about efficiency. It's just about flexibility and expressiveness.

Right now, you can't say things like: use my reservation first, and if not possible, do a ProvReq.

<!--
This section is for explicitly listing the motivation, goals, and non-goals of
this KEP.  Describe why the change is important and the benefits to users. The
motivation section can optionally provide links to [experience reports] to
demonstrate the interest in a KEP within the wider Kubernetes community.

[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
-->

### Goals

> - Support running AdmissionChecks only when the ResourceFlavor is used.

Reviewer 2: ```suggestion
- Support running AdmissionChecks only when a particular ResourceFlavor is assigned to the Workload.
```

> - Reduce the number of unnecessary AdmissionChecks.

Reviewer 2: I don't think this should be a goal.

What about running different an AdmissionCheck with different settings depending on the flavor? That could be a goal.

<!--
List the specific goals of the KEP. What is it trying to achieve? How will we
know that this has succeeded?
-->

### Non-Goals

<!--
What is out of scope for this KEP? Listing non-goals helps to focus discussion
and make progress.
-->

## Proposal

> Add new field to FlavorQuota API, so it contains list of AdmissionChecks thatshould run when the ResourceFlavor is used.

Reviewer 2: ```suggestion
Add new field to FlavorQuota API, so it contains list of AdmissionChecks thatshould run when a Workload
is admited for this ResourceFlavor.
```

<!--
This is where we get down to the specifics of what the proposal actually is.
This should have enough detail that reviewers can understand exactly what
you're proposing, but should not include things like API designs or
implementation. What is the desired outcome and how do we measure success?.
The "Design Details" section below is for the real
nitty-gritty.
-->

### User Stories (Optional)

<!--
Detail the things that people will be able to do if this KEP is implemented.
Include as much detail as possible so that people can understand the "how" of
the system. The goal here is to make this feel real for users without getting
bogged down.
-->

#### Story 1
As a user who has reserved machines at my cloud provider I would like to use them first. If they are not obtainable then
switch to spot machines, using ProvisioningRequest. It only makes sense to run Provisioning AdmissionCheck when the ResourceFlavor is Spot.

### Notes/Constraints/Caveats (Optional)

> All `AdmissionChecks` assigned to a Workload must have different controllers.

Reviewer 2: Why? Please expand.

Will this be validated?

-->

### Risks and Mitigations

<!--
What are the risks of this proposal, and how do we mitigate? Think broadly.
For example, consider both security and how this will impact the larger
Kubernetes ecosystem.

How will security be reviewed, and by whom?

How will UX be reviewed, and by whom?

Consider including folks who also work outside the SIG or subproject.
-->

## Design Details
We propose adding a new field to `FlavorQuota` API called `AdmissionChecks` in following manner:

```
type FlavorQuotas struct {
  // name of this flavor. The name should match the .metadata.name of a
  // ResourceFlavor. If a matching ResourceFlavor does not exist, the
  // ClusterQueue will have an Active condition set to False.
  Name ResourceFlavorReference `json:"name"`  

  // resources is the list of quotas for this flavor per resource.
  // There could be up to 16 resources.
  // +listType=map
  // +listMapKey=name
  // +kubebuilder:validation:MinItems=1
  // +kubebuilder:validation:MaxItems=16
  Resources []ResourceQuota `json:"resources"`

  // admissionChecks is the list of AdmissionChecks that should run when the ResourceFlavor is used.

>   // No two AdmissionChecks should have the same controller.

Reviewer 2: what happens if they have the same controller? Is the workload not able to be admitted, is it marked as Failed somehow?

Author: WDYT about introducing a new condition and setting it if this happens?

  // +listType=set
  AdmissionChecks []string `json:"admissionchecks"`
}
```

> At the same time, we want to preserve the existing `AdmissionChecks` field in `ClusterQueue` API. A Workload should have

Reviewer: @mwielgus Apart from maintaining backwards compatibility (which I don't want to discount, though AdmissionChecks are still relatively young), is there a good reason to have AdmissionChecks at both the CQ and Flavor level?

I suppose not having to reason about AdmissionCheck override makes the API simpler to understand, and easier to debug when something is misbehaving

Reviewer 2: Even though it's new, it's already being used in production, so I definitely would leave it supported.

Reviewer 3: Right, but we still could consider allowing only for one level of AdmissionChecks - either CQ or RF, as suggested also in https://github.com/kubernetes-sigs/kueue/pull/1935#discussion_r1549503754

Author: Maybe we could simplify it and skip overriding as there are not that many use cases behind it. However, I would keep having AdmissionChecks both at the CQ and Flavor level, and I've added a Story to show an example. WDYT?

Author: I agree with @gabesaba and @mimowo https://github.com/kubernetes-sigs/kueue/pull/1935#discussion_r1551617863

existing tests to make this code solid enough prior to committing the changes necessary
to implement this enhancement.

<!-- ##### Prerequisite testing updates -->

<!--
Based on reviewers feedback describe what additional tests need to be added prior
implementing this enhancement to ensure the enhancements have also solid foundations.
-->

#### Unit Tests

Test Workload's Controller's function that assigns AdmissionChecks to a Workload
<!--
In principle every added code should have complete unit test coverage, so providing
the exact set of tests will not bring additional value.
However, if complete unit test coverage is not possible, explain the reason of it
together with explanation why this is acceptable.
-->

<!--
Additionally, try to enumerate the core package you will be touching
to implement this enhancement and provide the current unit coverage for those
in the form of:
- <package>: <date> - <current test coverage>

This can inform certain test coverage improvements that we want to do before
extending the production code to implement this enhancement.
-->

<!-- - `<package>`: `<date>` - `<test coverage>` -->

#### Integration tests

> - Create 2 ResourceFlavors with associated 2 AdmissionCheck and test if a Workload contains the AdmissionCheck associated with one of the ResourceFlavors;

Reviewer 2: ```suggestion
- Create 2 ResourceFlavors each with different AdmissionChecks and test if a Workload contains the AdmissionCheck associated with one of the ResourceFlavors;
```
?

-->

## Motivation

Currently, AdmissionChecks are only assigned to a ClusterQueue. This results in running `AdmissionChecks` regardless of
the ResourceFlavor used by a Workload, even when it is unnecessary. We would like to have the ability to assign AdmissionChecks
to a specific ResourceFlavor, so the whole mechanism is more flexible and efficient.

<!--
This section is for explicitly listing the motivation, goals, and non-goals of
this KEP.  Describe why the change is important and the benefits to users. The
motivation section can optionally provide links to [experience reports] to
demonstrate the interest in a KEP within the wider Kubernetes community.

[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
-->

### Goals
- Support running AdmissionChecks only when a particular ResourceFlavor is assigned to the Workload.
- Reduce the number of unnecessary AdmissionChecks.

<!--
List the specific goals of the KEP. What is it trying to achieve? How will we
know that this has succeeded?
-->

### Non-Goals

<!--
What is out of scope for this KEP? Listing non-goals helps to focus discussion
and make progress.
-->

## Proposal

> Add new field to FlavorQuota API, so it contains list of AdmissionChecks thatshould run when a Workload

Reviewer 3: ```suggestion
Add new field to FlavorQuota API, so it contains a list of AdmissionChecks that should run when a Workload
```

  // name of this flavor. The name should match the .metadata.name of a
  // ResourceFlavor. If a matching ResourceFlavor does not exist, the
  // ClusterQueue will have an Active condition set to False.
  Name ResourceFlavorReference `json:"name"`  

  // resources is the list of quotas for this flavor per resource.
  // There could be up to 16 resources.
  // +listType=map
  // +listMapKey=name
  // +kubebuilder:validation:MinItems=1
  // +kubebuilder:validation:MaxItems=16
  Resources []ResourceQuota `json:"resources"`

  // admissionChecks is the list of AdmissionChecks that should run when the ResourceFlavor is used.
  // No two AdmissionChecks should have the same controller.
  // +listType=set
  AdmissionChecks []string `json:"admissionchecks"`
}
```

At the same time, we want to preserve the existing `AdmissionChecks` field in `ClusterQueue` API. A Workload may have
assigned `AdmissionChecks` both from `ClusterQueue` and `FlavorQuota` APIs. In case two `AdmissionChecks` have the same
controller, the one from `FlavorQuota` API should take over.

This enables having default `AdmissionChecks` that apply to all Workloads submitted to a `ClusterQueue`, and more

> specific ones that apply only to Workloads using a specific `ResourceFlavor`.

Reviewer 3: Do we have a good story for supporting both levels at the same time? If so please add to the stories section. Otherwise I would suggest to simplify and only allow them to be present at one level. We can relax this in the future if needed. 

It feels that setting admission checks at both levels might complicate debugging for the "why my workload is not starting?" question.

Author: I've added a story for supporting this. Maybe we could simplify the mechanism though and discard overriding AdmissionChecks. WDYT

Reviewer 3: If the story is not yet motivated by a user request, then I would suggest to simplify. 

In case we have ACs only on one level we can apply the same validation rules, but scoped.

If we allow setting the Admissionchecks at two levels the semantic of merging and validation gets tricky. Some questions to emphasize the complications: 
1. why override, if we allow some admission checks (like ProvReq) to have multiple instances we could append rather than override,
2, why override rather than validation error for admission checks that require single instance, say for MK, user might get surprised which one "wins"
3. are the answers above different if the ACs have different / the same name?

Author: Thanks for clarifying. I took that into consideration and overall agree with it. I've committed changes that aligns with that approach

- [Summary](#summary)
- [Motivation](#motivation)
  - [Goals](#goals)
  - [Non-Goals](#non-goals)
- [Proposal](#proposal)
  - [User Stories (Optional)](#user-stories-optional)
    - [Story 1](#story-1)
  - [Notes/Constraints/Caveats (Optional)](#notesconstraintscaveats-optional)
  - [Risks and Mitigations](#risks-and-mitigations)
- [Design Details](#design-details)
  - [Test Plan](#test-plan)
    - [Unit Tests](#unit-tests)
    - [Integration tests](#integration-tests)
- [Drawbacks](#drawbacks)
- [Alternatives](#alternatives)
<!-- /toc -->

## Summary

This KEP introduces assigning AdmissionChecks to a ResourceFlavor.

## Motivation

Currently, AdmissionChecks are only assigned to a ClusterQueue. This results in running `AdmissionChecks` regardless of
the ResourceFlavor used by a Workload, even when it is unnecessary. We would like to have the ability to assign AdmissionChecks
to a specific ResourceFlavor, so the whole mechanism is more flexible and expressive.

### Goals
- Support running AdmissionChecks only when a particular ResourceFlavor is assigned to the Workload.
- Running an AdmissionCheck with different settings depending on the flavor

### Non-Goals

## Proposal

> Add a new field to FlavorQuota API, so it contains a list of AdmissionChecks that  should run when a Workload

Reviewer 4: ```suggestion
Add a new field to FlavorQuota API, so it contains a list of AdmissionChecks that should run when a Workload
```

is admitted for this ResourceFlavor.

### User Stories (Optional)

#### Story 1
As a user who has reserved machines at my cloud provider I would like to use them first. If they are not obtainable then
switch to spot machines, using ProvisioningRequest. It only makes sense to run Provisioning AdmissionCheck when the ResourceFlavor is Spot.

### Notes/Constraints/Caveats (Optional)
User cannot define AdmissionChecks both at ClusterQueue and ResourceFlavor level. This will be validated by
a ClusterQueue webhook.

> ### Risks and Mitigations

Reviewer 4: Are there any risks? If there are no risks, should we remove this section?

> ## Design Details

Reviewer 4: I think that this API design would bring AdmissionChecks fragmentation.
So, I'm wondering if we can add configurations under the `ClusterQueueSpec` field like this.
In theory, once we introduce the `AdmissionCheckStrategy`, the existing `AdmissionChecks` field is not needed, but we keep the existing field for backward compatibility.

@alculquicondor @PBundyra WDYT?

```diff
type ClusterQueueSpec struct {
...
    AdmissionChecks []string
+   AdmissionCheckStrategy []AdmissionCheckStrategy
}

+type AdmissionCheckPolicy struct {
+    // AdmissionCheck name.
+    Name string
+    Strategy AdmissionCheckStrategyKind
+    Selector  *metav1.LabelSelector OR []types.NamespacedName
+}

+type AdmissionCheckStrategyKind string
+constants (
+    AdmissionCheckStrategyKindClusterQueue AdmissionCheckStrategyKind = "ClusterQueue" // default
+    AdmissionCheckStrategyKindResourceFlavor AdmissionCheckStrategyKind = "ResourceFlavor"
+)
+
```

Reviewer 2: What is the label selector for? Is it for workload labels?

I'm failing to see what would be the use case at the moment, but I agree that we might need to add more "options" in the future.

OTOH, just a `strategy` is not enough. Maybe it could be something like:

```diff
type ClusterQueueSpec struct {
...
    AdmissionChecks []string
+   AdmissionCheckStrategy []AdmissionCheckStrategy
}

+type AdmissionCheckPolicy struct {
+    // AdmissionCheck name.
+    Name string
+    ForFlavors []string
+}
```

And if `forFlavors` is empty, then the admission check applies to all of the ClusterQueue.

Reviewer 2: The advantage of the above API, compared to the original proposal, is that it's more future proof.

Reviewer 4: > What is the label selector for? Is it for workload labels?

It is for ResourceFlavor, but we can use it for Workloads in the future.
But I don't have a strong point of view against LabelSelector, and I just wanted to express somehow generic filters.

> I'm failing to see what would be the use case at the moment, but I agree that we might need to add more "options" in the future.

Yes, that's right. That is the main motivation for my proposed API.

> OTOH, just a strategy is not enough. Maybe it could be something like:
And if forFlavors is empty, then the admission check applies to all of the ClusterQueue.

I like your proposed API rather than my proposal, but the `ForFlavors` sounds slightly weired.
How about the `CheckForFlavors` or `ActionForFlavors`?

Also, should we limit the number of `forFlavors`? I guess it is possible to set `16`.
WDYT?

Reviewer 2: How about this?
```yaml
admissionCheckStrategies:
- name: foo
  onFlavors: [bar]
```

I think more words is just redundant.

Reviewer 4: it looks great! Thank you!

#### Unit Tests

- Test Workload's Controller's function that assigns AdmissionChecks to a Workload

- Test Cache that checks ClusterQueue's status

#### Integration tests

> - Create 2 ResourceFlavors each with a different AdmissionCheck and test if a Workload contains the AdmissionCheck associated with one of the ResourceFlavors;

Reviewer 3: nit: I think it makes sense to add the baseline scenario, for the use case we have: no AC for reservation flavor, and one AC for the spot/on-demand flavor.

Author: Added
